x-pack/plugins/inference/server/routes/chat_complete.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { schema, Type } from '@kbn/config-schema';
+import type { CoreSetup, IRouter, RequestHandlerContext } from '@kbn/core/server';
+import { isObservable } from 'rxjs';
+import { MessageRole } from '../../common/chat_complete';
+import type { ChatCompleteRequestBody } from '../../common/chat_complete/request';
+import { ToolCall, ToolChoiceType } from '../../common/chat_complete/tools';
+import { createInferenceClient } from '../inference_client';
+import { InferenceServerStart, InferenceStartDependencies } from '../types';
+import { observableIntoEventSourceStream } from '../util/observable_into_event_source_stream';
+
+const toolCallSchema: Type<ToolCall[]> = schema.arrayOf(
+  schema.object({
+    toolCallId: schema.string(),
+    function: schema.object({
+      name: schema.string(),
+      arguments: schema.maybe(schema.object({}, { unknowns: 'allow' })),
+    }),
+  })
+);
+
+const chatCompleteBodySchema: Type<ChatCompleteRequestBody> = schema.object({
+  connectorId: schema.string(),
+  system: schema.maybe(schema.string()),
+  tools: schema.maybe(
+    schema.recordOf(
+      schema.string(),
+      schema.object({
+        description: schema.string(),
+        schema: schema.maybe(
+          schema.object({
+            type: schema.literal('object'),
+            properties: schema.recordOf(schema.string(), schema.any()),
+            required: schema.maybe(schema.arrayOf(schema.string())),
+          })
+        ),
+      })
+    )
+  ),
+  toolChoice: schema.maybe(
+    schema.oneOf([
+      schema.literal(ToolChoiceType.auto),
+      schema.literal(ToolChoiceType.none),
+      schema.literal(ToolChoiceType.required),
+      schema.object({
+        function: schema.string(),
+      }),
+    ])
+  ),
+  messages: schema.arrayOf(
+    schema.oneOf([
+      schema.object({
+        role: schema.literal(MessageRole.Assistant),
+        content: schema.string(),
+        toolCalls: toolCallSchema,
+      }),
+      schema.object({
+        role: schema.literal(MessageRole.User),
+        content: schema.string(),
+        name: schema.maybe(schema.string()),
+      }),
+      schema.object({
+        role: schema.literal(MessageRole.Tool),
+        toolCallId: schema.string(),
+        response: schema.object({}, { unknowns: 'allow' }),
+      }),
+    ])
+  ),
+});
+
+export function registerChatCompleteRoute({
+  coreSetup,
+  router,
+}: {
+  coreSetup: CoreSetup<InferenceStartDependencies, InferenceServerStart>;
+  router: IRouter<RequestHandlerContext>;
+}) {
+  router.post(
+    {
+      path: '/internal/inference/chat_complete',
```

Comment:
nit: would be nice to have some `jest:integration` tests for these new endpoints... Although I'm not sure how possible that is unless we mock AI connector returns. Up to you.

Reply:
happy to add them, do you have any examples? I've only used FTR for API tests.

Reply:
Here is one I worked on recently: `src/core/server/integration_tests/http/oas.test.ts`. FTR tests are also a good option, I just tend to prefer the `jest:test_integration` as they are faster.